### DISCUSSION TITLE
IndexedDb Persister

### DISCUSSION THREAD

**Discussion by jonathanstanley** [TanStack/query#3198] [Ideas] [answered]
Now that we have [better persister in react-query v4 alpha](https://github.com/tannerlinsley/react-query/releases/tag/v4.0.0-alpha.7) 🚀, I was also going to see about sharing my IndexedDb persister.  It's super easy:
```javascript
import { get, set, del } from "idb-keyval";
import { PersistedClient, Persister } from "react-query/persistQueryClient";

/**
 * Creates an Indexed DB persister
 * @see https://react-query.tanstack.com/plugins/persistQueryClient#building-a-persistor
 * @see https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API
 */
export function createIDBPersister(idbValidKey: IDBValidKey = "reactQuery") {
  return {
    persistClient: async (client: PersistedClient) => {
      set(idbValidKey, client);
    },
    restoreClient: async () => {
      return await get<PersistedClient>(idbValidKey);
    },
    removeClient: async () => {
      await del(idbValidKey);
    },
  } as Persister;
}

```

# Benefits

1. In my use, I haven't seen it even come close to needing throttling.  [Indexed-DB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) is high performance and async.
2. There is no need to serialize/deserialize since Indexed DB supports the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).  dates, recursive objects, and other native types are all handled natively.  see also https://github.com/tannerlinsley/react-query/discussions/1638
3. no 5MB limit
4. ~~It possibly can replace WebStoragePersister (other than the dependency mentioned below, I can't think of a reason to use WebStoragePersister over this)~~ ([nevermind on this one; see discussion](https://github.com/tannerlinsley/react-query/discussions/3198#discussioncomment-1972926))

# Shall I submit as a PR?
I can also add documentation and examples for use. 

The only issue is dependencies.  This requires [idb-keyval](https://www.npmjs.com/package/idb-keyval) and `react-query` currently only has 3 dependencies.  I checked into implementing without a dependency, but `idb-keyval` handles a couple browser quirks.

**Comment by cliffordfajardo** [comment]
Nice!  Using the vanilla IndexDB API's are not a great experience so glad that IDB keyval exists; it definitely makes the implementation easier.

If [idb-keyval](https://www.npmjs.com/package/idb-keyval) is used for the IndexDB Persister, it should be included as a peerDependency like how `reactQuery` lists `react` as a [peerDependency here] ?(https://github.com/tannerlinsley/react-query/blob/master/package.json#L72) 


> It possibly can replace WebStoragePersister (other than the dependency mentioned below, I can't think of a reason to use WebStoragePersister over this)

I think having the option to use `LocalStorage` still might be good to keep for wider browser support? 🤔 

> no 5MB limit

This is no limit is nice. I've hit the limit 5MB limit very easily with localStorage, due to 3rd party libraries stuffing data in the localStorage for their own requirements, leaving less space for my app data
For example:
-  [AWS Amplify stores it's authentication data in LocalStorage](https://github.com/aws-amplify/amplify-js/search?q=localstorage)


## Some follow up questions are:
Are there any downsides/alternatives to including `idb-keyval`?
-  I really couldn't think of any off the top of my head

**Reply by jonathanstanley** [reply]
Thanks for your thoughts!  I generally find libraries with few or no dependencies to be much better (part of why I chose react-query).  Although idb-keyval is great, it isn't really to do with react-query... just this plugin.  So, for me, this is itself a downside to including it.  A peer-dependency wouldn't inflate the dependency count.  But a peer-dependency could actually be more annoying for users since they would be prompted to install idb-keyval in their package even if they don't use this IndexedDbPersister. In my mind, it doesn't seem right to set this as a peer dependency.

Your point about AWS is well taken.  I could see use cases that need to work with other systems that makes it preferable to provide persisters for each case.  I'll edit my OP.

re: browser support 
local storage does have a 1.8% advantage:
https://caniuse.com/indexeddb2 (93.4% of users have it)
https://caniuse.com/mdn-api_window_localstorage (95.2% of users have it)

**Comment by TkDodo** [comment]
isn't this just an async storage persister with a bit of renaming?

```
import { get as getItem, set as setItem, del as removeItem } from "idb-keyval";

createAyncStoragePersister({
  storage: { getItem, setItem, removeItem }
})
```

you would also profit from throttling. While we've made the `AsyncStorage` interface to work best with react-native's AsyncStorage, the interface is still pretty generic so that it should also work with things like `idb-keyval` or [localForage](https://github.com/localForage/localForage).

this is the interface that an `AsyncStorage` needs to fulfil: 

https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/createAsyncStoragePersister/index.ts#L3-L7

**Reply by jonathanstanley** [reply]
I believe you are right that it could be built with `asyncStoragePersister`, but there is an issue from how you posted.

```
createAyncStoragePersister({
  storage: { getItem, setItem, removeItem }
})
```

this would impose JSON.stringify (which we wouldn't want to do with IDB).  the small differences in my post:

1. No serialization.  Using the AsyncStorage interface would require creating a serialize and deserialize function that simply return the content unchanged
2. No throttle.  Using AsyncStorage requires choosing a throttle time or accepting the default.

createAsyncStoragePersister looks like it might be a little different from when I first read it a few months ago.  From what I'm reading today, the following is how it should be done:

```javascript
import { get as getItem, set as setItem, del as removeItem } from "idb-keyval";

createAyncStoragePersister({
  storage: { getItem, setItem, removeItem },
  serialize: (data)=>data,
  deserialize: (data)=>data
})
```

**Reply by TkDodo** [reply]
I must've missed that. If you actually want no-throttling, I think you should be able to just pass in `0`. Right now, this would throttle every 0ms, but I think we can re-write to turn off throttling altogether if `0` is passed in if that matters.

yes, `serialize` and `deserialize` options were added to both `createWebStoragePersister` and `createAsyncStoragePersister` to give more flexibility.

what I think would be cool would be to document what kind of async persisters / storages you can build with it, in the docs. Since everyone brings their own dependencies, this would be the best option imo.

**Reply by jonathanstanley** [reply]
I like the idea of including it as docs instead... this should be easy enough for others to implement  from that example.  I could submit a PR to add "Examples" to this page https://github.com/tannerlinsley/react-query/blob/alpha/docs/src/pages/plugins/createWebStoragePersister.md ?  Or where in the docs would you like?

IMO, it is enough that users can set throttle to `0` if they don't want to use throttle.

**Reply by TkDodo** [reply]
yes, to the `alpha` branch would be perfect 👍 . but to the async storage persister, right: https://github.com/tannerlinsley/react-query/blob/alpha/docs/src/pages/plugins/createAsyncStoragePersister.md

> IMO, it is enough that users can set throttle to 0 if they don't want to use throttle.

The question is if the current implementation is enough, because it would basically do a `setTimeout(0)`, which is still some sort of deferring. 

we could, for example, _not_ call `asyncThrottle` at all when the `throttleTime` is zero:

https://github.com/tannerlinsley/react-query/blob/9a20452f7139de90a5d94021cfe57dbd87cc9256/src/createAsyncStoragePersister/index.ts#L37-L40

**Reply by jonathanstanley** [reply]
Sounds good.  I may be able to submit the PR later this week.  As far as the throttle... I have no preference.  In my case, I didn't need throttle, so I didn't build it.  I think `0` would work in my use, however, I don't know if others may want to actually disallow throttling.

**Reply by jonathanstanley** [reply]
the types are not compatible.  [serialize/deserialize expects a string](https://github.com/tannerlinsley/react-query/blob/293e5e9891e56f1b420475b5af4cb30b3a44acce/src/createAsyncStoragePersister/index.ts#L17-L26):

<img width="700" alt="Screen Shot 2022-02-16 at 10 24 04 AM" src="https://user-images.githubusercontent.com/2755958/154296941-dbe7d3aa-179c-4540-ad8c-dc53950f8600.png">

```
Type '(data: string) => string' is not assignable to type '(cachedString: string) => PersistedClient'.
  Type 'string' is not assignable to type 'PersistedClient'.ts(2322)
[index.d.ts(24, 5): ]()The expected type comes from property 'deserialize' which is declared here on type 'CreateAsyncStoragePersisterOptions'
```

similarly, the [getItem/setItem expects a string](https://github.com/tannerlinsley/react-query/blob/293e5e9891e56f1b420475b5af4cb30b3a44acce/src/createAsyncStoragePersister/index.ts#L4-L5).

**Reply by jonathanstanley** [reply]
1) i took a look at performance.  as best I can find, throttling wouldn't help with indexeddb.  rather, [google says](https://developers.google.com/web/fundamentals/instant-and-offline/web-storage/indexeddb-best-practices#keeping_your_app_performant) best practice is:
> Instead of storing the entire state tree in a single record, you should break it up into individual records and only update the records that actually change.

that suggests the better thing would be to cache individual keys instead of bundling it all in the `PersistedClient` .  conversely, [tests seem to show it is better to consolidate transactions](http://nparashuram.com/IndexedDB/perf/#Transaction%20based%20on%20Write).  Not sure which is correct... that is another topic. either way, throttling doesn't help because indexeddb isn't dom-blocking in itself.

2) indexeddb specifically does not want serialization

these are the two features of the `createAsyncStoragePersister`, so as I'm reading it now, it seems strange to try and use it.  better to just create a `Persister` directly as in my OP.  I could add that as an example to the docs instead?

**Reply by TkDodo** [reply]
> the types are not compatible

we can make the `AsyncPersister` interface be generic and make it infer the type from whatever `persist` returns, like:

```
interface AsyncStorage<TValue = string> {
  getItem: (key: string) => Promise<TValue | null>
  setItem: (key: string, value: TValue) => Promise<void>
  removeItem: (key: string) => Promise<void>
}

interface CreateAsyncStoragePersisterOptions<TValue = string> {
  storage: AsyncStorage<TValue>
  key?: string
  throttleTime?: number
  serialize?: (client: PersistedClient) => TValue
  deserialize?: (cachedString: TValue) => PersistedClient
}
```

> that suggests the better thing would be to cache individual keys instead of bundling it all in the PersistedClient 

we have a separate discussion for this:
- https://github.com/tannerlinsley/react-query/discussions/2649

but right now saving the whole client is the only option for persisters.

**Reply by TkDodo** [reply]
@jonathanstanley what do you think about the generic proposal?

**Reply by jonathanstanley** [reply]
Those interfaces look good and I think would help create persisters that don't serialize to strings. 👍

However, thinking about the bigger picture, I'm not sure it is appropriate to use for Indexed DB persister.  Since Indexed DB doesn't serialize and is async, I can't think of any purpose of a throttle.  Rather, even a throttle 0 may introduce some slight overhead and slow it down.  Without a throttle or serialization, the `createAsyncStoragePersister` doesn't add anything for Indexed DB persister.  So, it seems odd to use it. And using it as an example for `createAsyncStoragePersister` could be confusing for users.

FWIW, when I first looked at persisters i was discouraged from [building a persister](https://react-query.tanstack.com/plugins/persistQueryClient#building-a-persistor) and thought the two creator functions were the only way.  Perhaps the wording "Officially Supported Persisters" is a little confusing because they aren't actually persisters... they are _[builders](https://en.wikipedia.org/wiki/Builder_pattern)_ for persisters.  

In any case, I actually think Indexed DB persister would be a great candidate for a "Build your own persister example"  .  

1.  I'd be happy to submit a PR for this as an example [in the alpha docs](https://github.com/tannerlinsley/react-query/blob/alpha/docs/src/pages/plugins/persistQueryClient.md?plain=1#L166).
2. I could also propose some changes to help clear up the wording a bit.

**Reply by TkDodo** [reply]
> In any case, I actually think Indexed DB persister would be a great candidate for a "Build your own persister example" .

that would be awesome, thank you. Happy to accept contributions for that, and also 🟢  for clearing up any wordings you deem confusing 👍 

>  Rather, even a throttle 0 may introduce some slight overhead and slow it down.

My suggestion would be to _not_ call `throttleAsync` at all in the `AsyncPersister` if a value of `0` is passed. Or maybe a value of `false` to turn it off, widening the type to `number | false`. Thoughts?

**Reply by jonathanstanley** [reply]
A throttle 0 skipping the throttle code may help others? Or maybe it would just add complexity?  I'm really not sure what the use case would be, so I don't know that I have helpful feedback on that.

I'll send a PR soon.

**Reply by jonathanstanley** [reply]
submitted PR: https://github.com/tannerlinsley/react-query/pull/3356

**Comment by jonathanstanley** [comment, marked as answer]
merged in PR: https://github.com/tannerlinsley/react-query/pull/3356

**Reply by TkDodo** [reply]
@jonathanstanley do we still need to adapt the `AsyncPersister` to accept generics as [proposed here](https://github.com/tannerlinsley/react-query/discussions/3198#discussioncomment-2189865)?

**Reply by jonathanstanley** [reply]
It could be helpful for others developing persisters, however it isn't needed for my contributions.

**Reply by onekiloparsec** [reply]
@TkDodo does your proposition still holds? Like @CavalcanteLeo below, I am very much interested having such persister for `query-vue`. I am currently trying to write one using the dispatched information here and there for IndexedDB.

The little factory I've written (see below) uses the experimental function `experimental_createPersister`. It seems to work, but indeed goes through the (standard JSON) serialization. Should I open an issue? I am not yet familiar enough with the codebase for opening a PR.

**Comment by schachfeld** [comment]
Thank you for the insight! This is a way around the 5MB limit for `localStorage`.
I have one problem though: IndexedDB is really slow. I am writing about 300 MB of data which of course is pretty massive. The problem however is that the `AsyncPersister` saves the Cache on every url change (react router). I tried to use throttle already which had a positive impact on render time but when I reloaded the page some of the already cached requests where being fetched again (I have `staleTime` and `cacheTime` on Infinity right now).
Is there any solution to this? I really like this library but I need a way to cache this massive amount of data...

**Reply by TkDodo** [reply]
>  but when I reloaded the page some of the already cached requests where being fetched again

are you using v4 with the `PersistQueryClientProvider` ?

**Reply by jonathanstanley** [reply]
see https://github.com/TanStack/query/discussions/2649

**Reply by schachfeld** [reply]
> > but when I reloaded the page some of the already cached requests where being fetched again
> 
> are you using v4 with the `PersistQueryClientProvider` ?

yes



> see #2649

Thanks :) Persisting the queries individually seems like a solution


I am still wondering why the `persistQueryClient` persists every time I change the route rather than only if I am trying to exit the site.

**Reply by TkDodo** [reply]
Show the behavior in a sandbox please

**Reply by nhd2106** [reply]
@schachfeld Hi, did you resolve your problem, sir?

**Reply by greatstudentkaze** [reply]
@nhd2106 hi! did you manage to solve this problem?

**Reply by nhd2106** [reply]
Hi, just update your all tanstack-query dependencies to the same version.

**Reply by greatstudentkaze** [reply]
@nhd2106 all deps already have the same 4.20.4 version.. or you mean to v5?

![CleanShot 2024-03-26 at 09 40 15@2x](https://github.com/TanStack/query/assets/51622226/f1600717-b6c4-4f85-afff-ccccbed0edc8)

**Reply by dberardo-com** [reply]
is this still an issue ? i could not understand well the problem here so i dont know whether that affects me, but would like to be on the safe side

**Comment by milahu** [comment]
feature request:
all queries should wait for `persister.restoreClient`

in my case, restoreClient takes 0.5 seconds,
but my interface is faster, so queries are fetched, not loaded from cache

related: XY problem issue in #4368

this would help: a global setting `queryClient.enabled` so we can disable all queries
or `queryClient.paused` as a "default undefined" option

draft ...

<details>

`node_modules/@tanstack/query-core/build/lib/queryObserver.esm.js`

```js
  onSubscribe() {
    if (this.listeners.length === 1) {
      this.currentQuery.addObserver(this);

      if (shouldFetchOnMount(this.currentQuery, this.options)) {

        //this.executeFetch();

        const queryClient = this.client
        if (!queryClient.paused) {
          this.executeFetch();
        }
        else {
          // queryClient is paused -> fetch later
          if (!queryClient.afterPersistQueryClientRestore) {
            queryClient.afterPersistQueryClientRestore = []
          }
          queryClient.afterPersistQueryClientRestore.push(() => {
            if (shouldFetchOnMount(this.currentQuery, this.options)) {
              this.executeFetch();
            }
          })
        }

      }

      this.updateTimers();
    }
  }
```

`node_modules/@tanstack/query-persist-client-core/build/lib/persist.esm.js`

```js
async function persistQueryClientRestore({
  queryClient,
  persister,
  maxAge = 1000 * 60 * 60 * 24,
  buster = '',
  hydrateOptions
}) {
  queryClient.paused = true

  try {
    // ...
  } catch (err) {
    // ...
  }

  queryClient.paused = false
  if (queryClient.afterPersistQueryClientRestore) {
    for (const fn of queryClient.afterPersistQueryClientRestore) {
      fn()
    }
  }
}
```

</details>

**Reply by TkDodo** [reply]
> feature request:
all queries should wait for persister.restoreClient

again, that is what the `PersistQueryClientProvider` is doing. It's documented:

https://tanstack.com/query/v4/docs/plugins/persistQueryClient#usage-with-react
